### PR TITLE
[Serializer] Feature deserialization of object with private constructor

### PR DESCRIPTION
**Pull request by l3l0**
| Q | A |
| --- | --- |
| Branch? | "master" |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR |  |

See: https://github.com/symfony/symfony/pull/19025

### CODE REVIEW

**Comment by dunglas** [comment]
Can you describe a bit more the use case?

File: src/Symfony/Component/Serializer/Tests/SerializerTest.php [hunk 1]
```diff
@@ -195,6 +195,14 @@ public function testDeserialize()
         $this->assertEquals($data, $result->toArray());
     }
 
+    public function testDeserializeObjectWithPrivateConstructor()
+    {
+        $serializer = new Serializer(array(new GetSetMethodNormalizer()), array('json' => new JsonEncoder()));
+        $data = array('title' => 'foo', 'numbers' => array(5, 3));
+        $result = $serializer->deserialize(json_encode($data), '\Symfony\Component\Serializer\Tests\ModelWithPrivateConstructor', 'json');
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Tests/SerializerTest.php]: Can you use the ::class constant instead of the FQCN as string?

**Comment by Ener-Getick** [comment]
@dunglas this can be useful when you use a third party library allowing to instantiate its classes with static methods (even if this proposal looks risky to me and i would prefer to be able to customize the object factory).

**Comment by dunglas** [comment]
IMO it would be easier to use a custom normalizer for that kind of usage.

**Comment by Ener-Getick** [comment]
@dunglas I agree even if i think we should allow to simply override the class constructor with something like:
```php
protected function getConstructor(string $class, \ReflectionClass $reflectionClass) : \ReflectionMethod
{
    return $reflection->getConstructor();
}

protected function instantiateObject(/** ... */)
{
    // ...
    $constructor = $this->getConstructor($class, $reflectionClass);
    // ...
    if ($constructor->isConstructor()) {
        return $reflectionClass->newInstanceWithArgs($args);
    } else {
        return $constructor->invokeArgs(null, $args);
    }
}
```
WDYT ?

@l3l0 would also be able to fix his issue without introducing a hack in the core.

**Comment by dunglas** [comment]
@Ener-Getick Good idea!

**Comment by l3l0** [comment]
@dunglas @Ener-Getick  use case is quite simple... serialize and unserialize events with private constructor (cause events have many [named constructor](http://verraes.net/2014/06/named-constructors-in-php/)) and I do not want to break encapsulation and make constructor public only because serializer do not handle that :) This can be the problem with serializing other kind of Value Objects as well.
I resolved my issue by writing custom normalizer for now. 

Extension point proposed by @Ener-Getick seems to make such changes easier and it seems like good idea (I would even thiking about delegating instantiateObject to separate object/class).

We should consider change default behavior of property normalizer as well, cause it seems strange to me that data are read/write by reflection and calling constructor is required (or at least it should be optional)

**Comment by Ener-Getick** [comment]
@l3l0 i don't think this is worth creating a new class. At worst, we can allow to pass a callable to the normalizer as what is done to manage circular references but i think the right way is to extend it.

**Comment by fabpot** [comment]
I like @Ener-Getick proposal more than the current implementation of this PR.

**Comment by Ener-Getick** [comment]
I opened https://github.com/symfony/symfony/pull/19137.

**Pull request closed without merging**